### PR TITLE
Implement function type variance

### LINKED ISSUE #907: Implement function type descriptors and values
At the moment, the only thing you can do with an identifier of a function definition is use it in a function call expression.

In this stage, we allow functions to be values, which can be (like other values):
- stored in variables
- passed as parameters
- returned from functions
- be the result of expressions

In this stage, the only way to create a function value is with a reference to the identifier of a function definition. Anonymous functions would be supported in #908.

The syntactic change is to support function type descriptors. (This is already implemented in semtypes.)

This also requires us to support casting a function value to another function type.

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
Implement function type descriptors and values
Related #907

### CODE REVIEW

File: compiler/modules/front/expr.bal [hunk 1]
```diff
@@ -1398,15 +1416,18 @@ function codeGenCheckingTerminator(bir:BasicBlock bb, s:CheckingKeyword checking
...
+    bir:AssignTmpRegister? funcRegister = ();
     if prefix == () {
         var ref = cx.lookupLocalVarRef(expr.funcName, expr.qNamePos);
         if ref is bir:FunctionRef {
             func = ref;
         }
         else if ref is Binding {
-            func = check functionRefFromRegister(cx, ref.reg, expr.qNamePos);
-            funcRegister = ref.reg;
+            var [semType, atom] = check selectFunctionInherentType(cx, ref.reg.semType, expr);
```
Comment (jclark) [inline compiler/modules/front/expr.bal]: I'm not sure this is the right semantics for function call. Have we looked at how the Covariance/Contravariance paper types functional call?  The spec doesn't have any concept of function inherent type.

Comment (heshanpadmasiri) [reply compiler/modules/front/expr.bal]: The problem I am trying to solve here is 
```
import ballerina/io;
type F function(int) returns int;
type F2 function(int, int) returns int;

public function main() {
    F|F2 f = f1;
    if f is F {
        io:println(f(1));
    }
    else {
        io:println(f(1, 2)); // <- what is the type of f
    }
}

function f1(int i) returns int {
  return i;
}
```
Currently at the point I have marked we get the type as a `diff` of two types (union and `F`), but we need a single function type (at least I think we need one to generate code).

Comment (heshanpadmasiri) [reply compiler/modules/front/expr.bal]: Fixed

File: runtime/function.c [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    if (restArgCount >= INT64_MAX - fixedArgCount) {
+        // realistically only way this could happen is when you spread a list with a fixed number of arguments
+        _bal_panic_internal(PANIC_LIST_TOO_LONG);
+    }
+    int64_t argCount = fixedArgCount + restArgCount;
+    TaggedPtr *arr = (TaggedPtr*)_bal_alloc(sizeof(TaggedPtr) * argCount);
+    return arr;
+}
+
+void _bal_add_rest_args(TaggedPtr* arr, int64_t index, TaggedPtr restArgArray) {
```
Comment (jclark) [inline runtime/function.c]: Call startIndex or offset

Comment (heshanpadmasiri) [reply runtime/function.c]: fixed

File: runtime/function.c [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        }
+    }
+    return true;
+}
+
+bool _bal_is_exact(FunctionSignaturePtr signature, FunctionValuePtr value) {
+    return signature == value->signature;
+}
+
+TaggedPtr* _bal_create_uniform_arg_array(int64_t fixedArgCount, int64_t restArgCount) {
```
Comment (jclark) [inline runtime/function.c]: Space before `*`

File: runtime/function.c [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    // We are using uint64_t to avoid overflow in case of restArgCount close to INT64_MAX.
+    uint64_t argCount = fixedArgCount + restArgCount;
+    TaggedPtr *arr = (TaggedPtr*)_bal_alloc(sizeof(TaggedPtr) * argCount);
```
Comment (jclark) [inline runtime/function.c]: Space before *

File: runtime/function.c [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            paramBitSet = fstp->paramBitSets[i];
+        }
+        if (!memberTypeIsSubtypeSimple(paramTy, paramBitSet)) {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool _bal_is_exact(FunctionSignaturePtr signature, FunctionValuePtr value) {
```
Comment (manuranga) [inline runtime/function.c]: should this be named `_bal_function_is_exact`?

File: runtime/function.c (hunk 3 above)
Comment (heshanpadmasiri) [reply runtime/function.c]: fixed

File: runtime/function.c (hunk 4 above)
Comment (heshanpadmasiri) [reply runtime/function.c]: fixed

File: runtime/function.c (hunk 5 above)
Comment (heshanpadmasiri) [reply runtime/function.c]: fixed

File: compiler/modules/nback/basic.bal [hunk 6]
```diff
@@ -230,34 +230,122 @@ function buildCall(llvm:Builder builder, Scaffold scaffold, bir:CallInsn insn) r
...
-    RetRepr retRepr = semTypeRetRepr(signature.returnType);
-    buildStoreRet(builder, scaffold, retRepr, retValue, insn.result);
+    buildStoreRet(builder, scaffold, semTypeRetRepr(returnType), retValue, insn.result);
+    builder.br(afterCall);
 }
 
-function buildFunctionCallArgs(llvm:Builder builder, Scaffold scaffold, t:SemType[] paramTypes, t:SemType[] instantiatedParamTypes, bir:Operand[] args) returns llvm:Value[]|BuildError {
-    return from int i in 0 ..< args.length()
-           select check buildWideRepr(builder, scaffold, args[i], semTypeRepr(paramTypes[i]), instantiatedParamTypes[i]);
+function buildNotExactCall(llvm:Builder builder, Scaffold scaffold, bir:CallIndirectInsn insn,
```
Comment (manuranga) [inline compiler/modules/nback/basic.bal]: `Inexact` or `NonExact` sounds better to me.

**Comment by manuranga** [comment]
Do you have a test case where a function with no return being assigned to type that has an optional return (eg `int?`)?

File: compiler/modules/nback/basic.bal (hunk 6 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback/basic.bal]: Fixed

**Comment by heshanpadmasiri** [comment]
> Do you have a test case where a function with no return being assigned to type that has an optional return (eg `int?`)?

Added [`15-function/assign27-v.bal`](https://github.com/ballerina-platform/nballerina/blob/eb99410f9e1383112b6bd2cbdcf96557e834a71d/compiler/testSuite/15-function/assign27-v.bal)

File: compiler/modules/nback/basic.bal [hunk 7]
```diff
@@ -230,34 +230,122 @@ function buildCall(llvm:Builder builder, Scaffold scaffold, bir:CallInsn insn) r
...
+    llvm:BasicBlock afterCall = scaffold.addBasicBlock();
+    builder.condBr(isExact, ifExact, ifNotExact);
+    builder.positionAtEnd(ifExact);
+    check buildExactCall(builder, scaffold, insn, afterCall, funcStructPtr, signature);
+    builder.positionAtEnd(ifNotExact);
+    check buildInexactCall(builder, scaffold, insn, afterCall, funcStructPtr, signature);
+    builder.positionAtEnd(afterCall);
+}
+
+function buildExactCall(llvm:Builder builder, Scaffold scaffold, bir:CallIndirectInsn insn,
```
Comment (jclark) [inline compiler/modules/nback/basic.bal]: Can we factor out some commmon functionality between buildExactCall and buildCall?

buildExactCall doesn't feel quite the right name.

File: compiler/modules/nback/init.bal [hunk 8]
```diff
@@ -14,6 +14,11 @@ final llvm:StructType llIntRangeType = llvm:structType([LLVM_INT, LLVM_INT]);
 public type ProgramModule readonly & record {|
     bir:ModuleId id;
     TypeUsage typeUsage;
+    FunctionSignatureUsage functionUsage;
```
Comment (jclark) [inline compiler/modules/nback/init.bal]: Better to use another kind of TypeUsage.

Comment (heshanpadmasiri) [reply compiler/modules/nback/init.bal]: Fixed

File: compiler/modules/nback/basic.bal (hunk 7 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback/basic.bal]: Fixed

File: compiler/modules/nback/types.bal [hunk 9]
```diff
@@ -8,14 +8,31 @@ import wso2/nballerina.print.llvm;
 const USED_INHERENT_TYPE = 0x1;
 const USED_EXACTIFY = 0x2;
 const USED_TYPE_TEST = 0x4;
+const USED_FUNCTION_SIGNATURE_VALUE = 0x8;
```
Comment (jclark) [inline compiler/modules/nback/types.bal]: USED_INHERENT_TYPE => USED_CONSTRUCT
USED_FUNCTION_SIGNATURE_CALL => USED_CALLED

File: runtime/balrt.h [hunk 10]
```diff
@@ -347,7 +355,24 @@ typedef GC struct LargeString {
 } *LargeStringPtr;
 
 typedef void (*FunctionPtr)();
+typedef TaggedPtr (*UniformCallFunction)(TaggedPtr uniformArgs, int64_t nArgs, FunctionPtr funcPtr);
+
+typedef struct FunctionSignature {
+    UniformCallFunction uniformCallFunction;
+    MemberType returnTy;
```
Comment (jclark) [inline runtime/balrt.h]: Not consistent with rest of file.

Comment (heshanpadmasiri) [reply runtime/balrt.h]: Fixed

File: compiler/modules/nback/types.bal (hunk 9 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback/types.bal]: Fixed

File: runtime/balrt.h [hunk 11]
```diff
@@ -311,6 +311,14 @@ typedef struct {
     TaggedPtr strs[];
 } *StringSubtypePtr;
 
+typedef struct {
+    UniformSubtype uniform;
+    uint32_t returnBitSet;
+    uint32_t restBitSet;
+    int64_t nRequiredParams;
```
Comment (jclark) [inline runtime/balrt.h]: uint32_t

File: runtime/function.c [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+void _bal_add_uniform_args_to_rest_args(TaggedPtr restArgArray, const TaggedPtr *uniformArgArray, int64_t startingOffset, int64_t restArgCount) {
+    ListPtr lp = taggedToPtr(restArgArray);
+    for (int64_t i = 0; i < restArgCount; i++) {
+        uint64_t index = startingOffset + i;
+        TaggedPtr val = uniformArgArray[index];
+        PanicCode err = lp->desc->set(restArgArray, lp->gArray.length, val);
+        if (err != 0) {
+            _bal_panic_internal(err);
```
Comment (jclark) [inline runtime/function.c:54]: This doesn't look right.

File: runtime/function.c [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+void _bal_add_rest_args_to_uniform_args(TaggedPtr *uniformArgArray, const TaggedPtr restArgArray, int64_t startingOffset) {
+    ListPtr lp = taggedToPtr(restArgArray);
+    int64_t len = lp->tpArray.length;
+    for (int64_t i = 0; i < len; i++) {
+        uint64_t index = startingOffset + i;
+        uniformArgArray[index] = lp->desc->get(restArgArray, i);
+    }
+}
+
+void _bal_add_uniform_args_to_rest_args(TaggedPtr restArgArray, const TaggedPtr *uniformArgArray, int64_t startingOffset, int64_t restArgCount) {
```
Comment (jclark) [inline runtime/function.c]: _bal_function_*

File: runtime/balrt.h [hunk 14]
```diff
@@ -347,7 +355,20 @@ typedef GC struct LargeString {
...
+    UniformCallFunction uniformCallFunction;
+    MemberType returnType;
+    MemberType restType;
+    int64_t nRequiredParams;
+    MemberType paramTypes[];
+} FunctionDesc, *FunctionDescPtr;
+
 typedef GC struct FunctionValue {
+    FunctionDescPtr desc;
     FunctionPtr funcPtr;
```
Comment (jclark) [inline runtime/balrt.h]: Just func

File: runtime/balrt.h [hunk 15]
```diff
@@ -347,7 +355,20 @@ typedef GC struct LargeString {
 } *LargeStringPtr;
 
 typedef void (*FunctionPtr)();
+typedef TaggedPtr (*UniformCallFunction)(TaggedPtr uniformArgs, int64_t nArgs, FunctionPtr funcPtr);
```
Comment (jclark) [inline runtime/balrt.h]: UniformFunctionPtr


func not funcPtr

File: runtime/function.c (hunk 12 above)
Comment (heshanpadmasiri) [reply runtime/function.c:54]: Fixed. Now we are preallocating the restArgArray therefor we should never get a panic here

File: runtime/function.c (hunk 13 above)
Comment (heshanpadmasiri) [reply runtime/function.c]: Fixed

File: runtime/balrt.h (hunk 15 above)
Comment (heshanpadmasiri) [reply runtime/balrt.h]: Fixed

File: runtime/balrt.h (hunk 11 above)
Comment (heshanpadmasiri) [reply runtime/balrt.h]: Fixed

File: runtime/function.c (hunk 13 above)
Comment (heshanpadmasiri) [reply runtime/function.c:44]: Fixed

File: runtime/balrt.h (hunk 14 above)
Comment (heshanpadmasiri) [reply runtime/balrt.h]: Fixed

File: runtime/balrt.h (hunk 15 above)
Comment (heshanpadmasiri) [reply runtime/balrt.h:358]: Fixed

File: runtime/function.c [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+void _bal_function_add_uniform_args_to_rest_args(TaggedPtr restArgArray, const TaggedPtr *uniformArgArray, int64_t startingOffset, int64_t restArgCount) {
+    ListPtr lp = taggedToPtr(restArgArray);
+    for (int64_t i = 0; i < restArgCount; i++) {
+        uint64_t index = startingOffset + i;
+        TaggedPtr val = uniformArgArray[index];
+        PanicCode err = lp->desc->set(restArgArray, lp->gArray.length, val);
+        if (err != 0) {
+            // This should never happen since we preallocate the array.
```
Comment (jclark) [inline runtime/function.c]: and the values are of the expected type

File: runtime/function.c [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+bool _bal_function_is_exact(FunctionDescPtr desc, FunctionValuePtr value) {
+    return desc == value->desc;
+}
+
+// nArgs = requiredArgCount + restArgCount
+// We are using uint64_t to avoid overflow in case of restArgCount close to INT64_MAX. This means indexing uniform arg array
+// must also be done using uint64_t
+GC TaggedPtr *_bal_function_construct_uniform_arg_array(uint64_t nArgs) {
```
Comment (jclark) [inline runtime/function.c]: _bal_function_alloc_uniform_args

File: compiler/modules/nback/function.bal [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+    else if retRepr is IntRepr {
+        return buildUntagInt(builder, scaffold, returnVal);
+    }
+    else if retRepr is FloatRepr {
+        return buildUntagFloat(builder, scaffold, returnVal);
+    }
+    return returnVal;
+}
+
+function uniformRepr(t:SemType ty) returns TaggedRepr {
```
Comment (jclark) [inline compiler/modules/nback/function.bal]: This doesn't feel like it's function specific. There's similar functionality needed for structure members, which are usually tagged.

File: runtime/balrt.h [hunk 19]
```diff
@@ -311,6 +311,15 @@ typedef struct {
...
 } *StringSubtypePtr;
 
+// This only handles cases where function is both atomic and all parameter and return types are unions of basic types
+// For others we are using PrecomputedSubtypePtr
+typedef struct {
+    UniformSubtype uniform;
+    uint32_t returnBitSet;
+    uint32_t nRequiredParams;
+    uint32_t paramBitSets[];
+} *AtomicFunctionSubtypePtr;
```
Comment (jclark) [inline runtime/balrt.h]: Let's use Easy not Atomic

File: compiler/modules/nback/init.bal [hunk 20]
```diff
@@ -242,21 +266,146 @@ function createMappingDescType(t:Context tc, t:SemType semType) returns llvm:Str
...
+    t:SemType? restParamType = signature.restParamType;
+    llvm:ConstValue restParamTy = restParamType == () ? getMemberType(cx, t:NEVER) : getMemberType(cx, restParamType);
+    int nRequiredParams = requiredParamCount(signature);
+    llvm:ConstValue paramTys = cx.llContext().constArray(LLVM_MEMBER_TYPE,
+                                                         from int i in 0 ..< nRequiredParams select getMemberType(cx, signature.paramTypes[i]));
+    return cx.llContext().constStruct([constTid(cx, tid), uniformFunc, returnTy, restParamTy,
+                                       constInt(cx, nRequiredParams), paramTys]);
+}
+
+function createUniformFunction(InitModuleContext cx, t:FunctionSignature signature, llvm:Builder builder, string symbol) returns llvm:Function {
```
Comment (manuranga) [inline compiler/modules/nback/init.bal]: In other places in `nback` we have `llvm:Builder` followed by `context` in the arg list. Let's follow that.

File: runtime/function.c (hunk 16 above)
Comment (heshanpadmasiri) [reply runtime/function.c]: Fixed

File: runtime/function.c (hunk 17 above)
Comment (heshanpadmasiri) [reply runtime/function.c]: Fixed

File: compiler/modules/nback/function.bal (hunk 18 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback/function.bal]: Fixed

File: runtime/balrt.h (hunk 19 above)
Comment (heshanpadmasiri) [reply runtime/balrt.h]: Fixed

File: compiler/modules/nback/init.bal (hunk 20 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback/init.bal]: Fixed

File: compiler/modules/nback/init.bal [hunk 21]
```diff
@@ -142,81 +176,104 @@ function buildMain(bir:ModuleId entryModId, string userMainName, t:FunctionSigna
...
+    buildInitTypesForUsage(builder, cx, modules, USED_TYPE_TEST);
+    buildInitTypesForUsage(builder, cx, modules, USED_CALLED);
 }
 
-function buildInitTypesForUsage(InitModuleContext cx, ProgramModule[] modules, TypeHowUsed howUsed) {
+function buildInitTypesForUsage(llvm:Builder builder, InitModuleContext cx, ProgramModule[] modules, TypeHowUsed howUsed) {
     foreach var mod in modules {
         var { types, uses } = mod.typeUsage;
         bir:ModuleId id = mod.id;
+        cx.setCurrentModuleState({ typeUsageCount: types.length(), id});
```
Comment (manuranga) [inline compiler/modules/nback/init.bal]: space after `id`

File: compiler/modules/nback/init.bal [hunk 22]
```diff
@@ -49,6 +49,11 @@ type FillerDescDefn record {|
     readonly llvm:ConstPointerValue defn;
 |};
 
+type CurrentModuleState record {|
```
Comment (jclark) [inline compiler/modules/nback/init.bal]: Better to make this a class.

File: compiler/modules/nback/init.bal (hunk 21 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback/init.bal]: Fixed

File: compiler/modules/nback/init.bal (hunk 22 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback/init.bal]: Fixed

**Review by jclark — APPROVED** [review]

**Pull request merged**